keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1763 @@
+  allocation.
+- *Resource kubelet plugin*: a component which cooperates with kubelet to prepare
+  the usage of the resource on a node.
+
+The utility library will be developed outside of Kubernetes and does not have
+to be used by drivers, therefore it is not described further in this KEP.
+
+### Resource allocation flow
+
+The following diagram shows how resource allocation works for a resource that
+gets defined inline in a Pod. For a full definition of ResourceClass,
+ResourceClaim and ResourceClaimTemplate see the [API](#API) section below.
+
+Several of these operations may fail without changing the system state. They
+then must be retried until they succeed or something else changes in the
+system, like for example deleting objects. These additional state transitions
+are not shown for the sake of simplicity.
+
+![allocation](./allocation.png)
+
+The flow is similar for a ResourceClaim that gets created as a stand-alone
+object by the user. In that case, the Pod reference that ResourceClaim by
+name. The ResourceClaim does not get deleted at the end and can be reused by
+another Pod and/or used by multiple different Pods at the same time (if
+supported by the driver). The resource remains allocated as long as the
+ResourceClaim doesn't get deleted.
+
+### API
+
+ResourceClaim and ResourceClass are new built-in types in a new
+`cdi.k8s.io/v1alpha1` API group. This was chosen instead of using CRDs because
+core Kubernetes components must interact with them and installation of CRDs as
+part of cluster creation is an unsolved problem.
+
+The PodSpec gets extended. Types and structs referenced from PodSpec
+(ResourceClaimTemplate, ResourceClaimSpec) must be placed in `core.k8s.io/v1`
+to avoid a cyclic dependency because `cdi.k8s.io/v1alpha1` depends on
+NodeSelector from the core API.
+
+Secrets are not part of this API: if a resource driver needs secrets, for
+example to access its own backplane, then it can define custom parameters for
+those secrets and retrieve them directly from the apiserver. This works because
+drivers are expected to be written for Kubernetes.
+
+```
+// ResourceClass is used by administrators to influence how resources
+// are allocated.
+type ResourceClass struct {
+	metav1.TypeMeta
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta
+
+	// DriverName determines which resource driver is to be used for
+	// allocation of a ResourceClaim that uses this class.
+	//
+	// Resource drivers have a unique name in reverse domain order
+	// (acme.example.com).
+	DriverName string
+
+	// Parameters holds arbitrary values that will be available to the
+	// driver when allocating a resource that uses this class. The driver
+	// will be able to distinguish between parameters stored here and and
+	// those stored in ResourceClaimSpec. These parameters here can only be
+	// set by cluster administrators.
+	Parameters runtime.RawExtension
```

Comment:
i think crossplane XRs and gatewayAPI policies are both trying to solve the problem of "how do we not duplicate everything to tweak a global param?". Both do so by putting extra layer(s) between the cluster-global base resource and the end user params.

As I understand, with this API the parameters are either cluster-global or in the pod spec, which makes them difficult to reuse and access control.

Reply:
See my proposal in https://github.com/kubernetes/enhancements/pull/3064/files/edc6dcbaa1df1f140e1435734a90e31d9fc40257#r874497156